Weak reference concurrent pool

gregw opened the pull request:
The Pool is now a list of Holder instances, each with a WeakReference to an Entry and an optional hard reference.

gregw committed:
Experiment with a weakly referenced ConcurrentPool

The Pool is now a list of Holder instances, each with a WeakReference to an Entry and an optional hard reference.

gregw commented:
@sbordet @lorban What is wrong with this approach?

Once an entry is acquired, the pool only holds a weak reference to it.  Once it is released it creates a hard reference to keep it in GC.

I do not think there is a race, as before... what am I missing?

gregw committed:
Experiment with a weakly referenced ConcurrentPool

The Pool is now a list of Holder instances, each with a WeakReference to an Entry and an optional hard reference.
back link may be hard

gregw committed:
Experiment with a weakly referenced ConcurrentPool

The Pool is now a list of Holder instances, each with a WeakReference to an Entry and an optional hard reference.
free on acquire

gregw committed:
Experiment with a weakly referenced ConcurrentPool

The Pool is now a list of Holder instances, each with a WeakReference to an Entry and an optional hard reference.
free on acquire
removed ThreadLocal cache

gregw committed:
Experiment with a weakly referenced ConcurrentPool

The Pool is now a list of Holder instances, each with a WeakReference to an Entry and an optional hard reference.
free on acquire
added Leak test

gregw committed:
Experiment with a weakly referenced ConcurrentPool

The Pool is now a list of Holder instances, each with a WeakReference to an Entry and an optional hard reference.
free on acquire
added warnings

gregw committed:
Experiment with a weakly referenced ConcurrentPool

The Pool is now a list of Holder instances, each with a WeakReference to an Entry and an optional hard reference.
free on acquire
added warnings

gregw committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into experiment/jetty-12/WeakReference-ConcurrentPool

gregw committed:
Experiment with a weakly referenced ConcurrentPool

The Pool is now a list of Holder instances, each with a WeakReference to an Entry and an optional hard reference.
free on acquire
added warnings

gregw commented:
@sbordet @lorban OK it was not as simple as it looked at first.... but I think I found the race (could be acquired and freed before it was held after release).     I think I have fixed it with a spin test (scary).   It now works for multiplexed pools as well.   I removed the cache because I think it is an unnecessary complexity.

lorban committed:
Modify ArrayByteBufferPool and ConcurrentPool to make weak-referencing configurable

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

gregw committed:
Experiment with a weakly referenced ConcurrentPool

The Pool is now a list of Holder instances, each with a WeakReference to an Entry and an optional hard reference.
free on acquire
added warnings

gregw commented:
@lorban I reverted to use a weak pool by default (with system property to change default).  Whilst we are an experimental branch, we want to test it as much as possible.  Perhaps once we merge, we switch the default (and have a better way of setting it).
I'm also using concrete Holder classes rather than if(weak) tests.

lorban commented on jetty-core/jetty-util/src/test/java/org/eclipse/jetty/util/ConcurrentPoolTest.java:
```diff
@@ -611,4 +610,133 @@ public void testRandomStrategy()
+        System.gc();
+        pool.sweep();
+
+        assertThat(pool.size(), is(2));
```
This assertion is probably flaky as there is no guarantee that after `System.gc()` any actually leaked object will see its weak ref cleared. 

I think awaitility can help here:

```java
await().atMost(5, TimeUnit.SECONDS).until(() ->
{
    System.gc();
    pool.sweep();
    return pool.size();
}, is(2));
```

lorban commented:
@gregw that's cool, and I currently have no strong opinion about whether we should set the pool as weak by default or not. 

The reason why I made weak ref configurable is to measure the perf impact of this PR, so I want to compare in order:
 - mainline vs this PR with hard refs only
 - this PR with hard refs only vs this PR with weak refs
 - mainline vs this PR with weak refs
to make sure this doesn't regress perf in any meaningful way, as the concurrent pool really is perf critical.

I don't mind your modifications; as I can configure an `ArrayByteBufferPool` with weak or strong references at will the default doesn't impact my perf testing.

lorban commented:
Perf report (so far):

This change has no measurable impact, no matter if weak referencing is enabled or not. The JIT inlines the entire buffer acquisition code in a single blob and most of the time is spent, in order: iterating the list, CASing the BiInteger in tryAcquire, updating the ref counter of the RetainableByteBuffer.

Basically, the cost stayed where it was expected to be, so we can concentrate on making sure the feature works. :+1:

gregw committed:
Experiment with a weakly referenced ConcurrentPool

The Pool is now a list of Holder instances, each with a WeakReference to an Entry and an optional hard reference.
free on acquire
added warnings
no race in tests

gregw commented:
@lorban so I'm doing a risk analysis for this, and I think the fundamental approach is a lot safer than before, as we rely on less moving parts.
I don't think there is any risk if we keep a strong link accidentally, as that is just case we have today and we expect the application to release.
Ditto if we don't keep a strong link when we should, then we might cleanup an entry we did not need to... but only if the application doesn't have a strong link.  So it would just be an idle entry shrink... not harm.

The only problem I can see is the spin wait to see a string link before it is cleaned.  If we get that wrong, then we spin forever!   My concern is what if between one spin and the next, the entry is acquired, released, acquired and released?      Do we need a similar spin on the hold method, so that it waits to see the previous free?    That will prevent multiple acquire/release cycles from happening during a spin.... actually I think I will implement that now!

gregw committed:
Experiment with a weakly referenced ConcurrentPool

The Pool is now a list of Holder instances, each with a WeakReference to an Entry and an optional hard reference.
free on acquire
added warnings
no race in tests

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ConcurrentPool.java:
```diff
@@ -597,4 +667,100 @@ public String toString()
+    {
+        Pool.Entry<P> getEntry();
+
+        void hold();
```
How about renaming this method `harden()`?

sbordet replied:
I prefer `release()` as it matches where it is called from.

gregw replied:
@sbordet but it is not a release.  A release only reduces the multiplexed reference count and only if that goes to zero do we create a non-weak reference.   So it is called from release, but is not the same.
Maybe `lastRelease` or `fullyReleased` ?   I don't mind `harden`

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ConcurrentPool.java:
```diff
@@ -597,4 +667,100 @@ public String toString()
+
+        void hold();
+
+        void free();
```
How about renaming this method `weaken()`?

sbordet replied:
I prefer `acquire()` as it matches where it is called from.

gregw replied:
@sbordet again it is not an acquire.  perhaps `firstAcquire`?     I don't mind `weaken`

lorban reviewed (commented):
I think this logic is correct, and unless I missed something, the race condition that you fix with a spin loop in hoder can only happen under one cirmunstance: when a release makes the multiplex counter go to 0 at the same time an acquire moves it to 1.

In that case, `hold()` **must** execute first so the concurrent `free()` call stays in the spin loop to wait for that moment before doing its job.

But I see no reason that would make `hold()` spin as I can't see a race condition that would trigger this need.

Wether I'm right or wrong, these new states and their transitions should be very carefully documented.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ConcurrentPool.java:
```diff
@@ -597,4 +667,100 @@ public String toString()
+            }
+
+            // hold must only be called when we know the holder will is free.
+            while (_strong != null && !entry.isTerminated())
+                Thread.onSpinWait();
+            _strong = entry;
```
I think the while loop is not necessary, because there cannot be a case where 2 thread enter hold().
This is because free() does not return until hold() is finished, due to the spin loop in free().
So there can be only 1 thread entering hold().

Similarly but differently, there cannot be 2 threads entering free(), because it's only called when the multiplex moves 0->1, and you cannot have the sequence 0->1->0 until free() has returned, because it returns the entry onto which call release.
So free() has already returned by the time code can release it and acquire it again.

So the race is only between hold() and free(), but hold() must always happen before free(), so only free() must spin waiting for hold() to complete.

There is no need to do the other way around because it cannot happen: I have to return from free() in order to give out an entry onto which call release().

gregw replied:
I think you are right..... but 0.1% of me thinks there might be a chance.... however, at least I have proved we could add a spin wait here if it every proves to be necessary.
I'll remove for now and add a comment about why it is not necessary.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ConcurrentPool.java:
```diff
@@ -597,4 +667,100 @@ public String toString()
+            ConcurrentEntry<P> entry = _weak.get();
+            if (entry == null)
+            {
+                LOG.warn("LEAKED {}", this);
```
These should be removed, because if it happens it will print millions of warnings in logs.
Plus no information about the leak, so the log line would be kinda useless.
Yes we leak but we do not accumulate them.
If we want to discover leaks, they are better tracked at the higher level (e.g. ABBP.Tracking).

gregw replied:
I'll call a protected method that will do a debug for now, but can be extended if necessary.

gregw replied:
I've added debug mechanism that reports more info on a leaked entry.  Thoughts?

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ConcurrentPool.java:
```diff
@@ -597,4 +667,100 @@ public String toString()
         }
     }
+
+    public interface Holder<P>
```
Should be `private`.

gregw replied:
I've made protected for now, as it is exposed in my proposed `leaded(Holder<P>)` method for debugging.  But if the logging in that method is OK, then could all be private.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ConcurrentPool.java:
```diff
@@ -386,10 +434,17 @@ public static class ConcurrentEntry<E> implements Entry<E>
+            holder = pool.weak ? new WeakHolder<>(this) : new StrongHolder<>(this);
+        }
+
+        private Holder<E> getHolder()
```
I'd just reference the final field and remove the method.

gregw replied:
why?

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ConcurrentPool.java:
```diff
@@ -53,27 +54,20 @@ public class ConcurrentPool<P> implements Pool<P>, Dumpable
      */
     public static final int OPTIMAL_MAX_SIZE = 256;
 
+    private static final boolean STRONG_DEFAULT = Boolean.getBoolean(ConcurrentPool.class.getName() + ".STRONG_DEFAULT");
```
I prefer explicit configuration over JVM system property.

gregw replied:
yep, this is just a temporary thing for testing.    I'll put a TODO in.   Ultimately we will need to pick what the default is and then have properties and modules and xml updated to allow configuration.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java:
```diff
@@ -116,7 +120,23 @@ public ArrayByteBufferPool(int minCapacity, int factor, int maxCapacity, int max
+     * @param maxDirectMemory the max direct memory in bytes, -1 for unlimited memory or 0 to use default heuristic
+     * @param weakPool true if the pooled buffers should be weakly referenced upon acquisition, false otherwise
+     */
+    public ArrayByteBufferPool(int minCapacity, int factor, int maxCapacity, int maxBucketSize, long maxHeapMemory, long maxDirectMemory, boolean weakPool)
```
Don't really like the telescopic additional "weakPool" parameter -- seems an implementation detail that leaks many layers up, and if we change our mind it'll be part of the public APIs for a long time.

gregw replied:
@sbordet Isn't this at odds with your previous comment about preferring "explicit configuration".    How would you configure this if not by constructor parameter?    Should we just have a different classes for Weak pools?

sbordet replied:
@gregw the comments above are now out of date, and the `weakPool` parameter is never used, so please remove it.

gregw committed:
Experiment with a weakly referenced ConcurrentPool

The Pool is now a list of Holder instances, each with a WeakReference to an Entry and an optional hard reference.
free on acquire
added warnings

lorban commented:
@gregw @sbordet I think we now reached the point where we all agree that this idea is technically sound, will work, and is worth implementing.

Besides naming (I'll skip that subject), there are still two points of disagreement or uncertainties:
 - What should be done when a leak is detected to help report the problem?
 - How this new feature should be made configurable?

Here are my current positions:
 - When a leak is detected, we shouldn't do more than incrementing a counter exposed via JMX. The kind of information we can get from the `ConcurrentPool` is IMHO insufficient to track the source of a leak. At least, in the context of the retainable buffers pool where `acquire()`, `retain()` and `release()` calls all need to be tracked to understand where the cause of the leak might be.
 - I'm tempted by the idea that weak referencing shouldn't be made configurable at all, and `ConcurrentPool` should have this feature always on. My top reason for this preference is that `ConcurrentPool` already is a complex piece of lock-free code that we should keep as simple as possible.

Thoughts?

gregw commented:
> @gregw @sbordet I think we now reached the point where we all agree that this idea is technically sound, will work, and is worth implementing. 

+1

> Besides naming (I'll skip that subject), there are still two points of disagreement or uncertainties:

On naming I can live with any of hold/free, strengthen/weaken,  lastRelease/firstAcquire      but not release/acquire 

> * What should be done when a leak is detected to help report the problem?
> * How this new feature should be made configurable?
> 
> Here are my current positions:
> 
> * When a leak is detected, we shouldn't do more than incrementing a counter exposed via JMX. The kind of information we can get from the `ConcurrentPool` is IMHO insufficient to track the source of a leak. At least, in the context of the retainable buffers pool where `acquire()`, `retain()` and `release()` calls all need to be tracked to understand where the cause of the leak might be.

I think that is a reasonable default mechanism.

However, I do think we should at least include something like my DebugWeakHolder, so that extra info can be obtained without the need to change code.   However, I'd probably change it to make it a static determination of debug, to give the JIT every opportunity to optimize away virtual calls (does that work?).        But for now, I think we should make this debug mechanism private and not extensible, as once protected it is part of the API.


> * I'm tempted by the idea that weak referencing shouldn't be made configurable at all, and `ConcurrentPool` should have this feature always on. My top reason for this preference is that `ConcurrentPool` already is a complex piece of lock-free code that we should keep as simple as possible.

+1

So long as we provide alternate classes that do simple or null pooling, then it is best to make our main ConcurrentPool implement the best algorithm we know and to not try to be too configurable unless there is good reason.

gregw committed:
Merge branch 'jetty-12.0.x' into experiment/jetty-12/WeakReference-ConcurrentPool

gregw committed:
A weakly referenced ConcurrentPool

The Pool is now a list of Holder instances, each with a WeakReference to an Entry and an optional hard reference.
free on acquire
added warnings

gregw committed:
Merge branch 'jetty-12.0.x' into experiment/jetty-12/WeakReference-ConcurrentPool

lorban commented:
@gregw I don't think `DebugWeakHolder` adds much value, if any, so I don't think it's worth the extra complexity in `ConcurrentPool`. Better leave leak tracking to the layer above, like `ArrayByteBufferPool.Tracking` IMHO.

gregw committed:
Merge branch 'jetty-12.0.x' into experiment/jetty-12/WeakReference-ConcurrentPool

gregw committed:
A weakly referenced ConcurrentPool

The Pool is now a list of Holder instances, each with a WeakReference to an Entry and an optional hard reference.
free on acquire
added warnings

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java:
```diff
@@ -482,13 +503,13 @@ private static class RetainedBucket
         private final int _capacity;
 
-        private RetainedBucket(int capacity, int poolSize)
+        private RetainedBucket(int capacity, int poolSize, boolean weakPool)
```
`weakPool` not used.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java:
```diff
@@ -566,6 +587,11 @@ public Quadratic(int minCapacity, int maxCapacity, int maxBucketSize)
+            this(minCapacity, maxCapacity, maxBucketSize, maxHeapMemory, maxDirectMemory, false);
+        }
+
+        public Quadratic(int minCapacity, int maxCapacity, int maxBucketSize, long maxHeapMemory, long maxDirectMemory, boolean weakPool)
```
`weakPool` not used.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java:
```diff
@@ -70,7 +70,8 @@ public class ArrayByteBufferPool implements ByteBufferPool, Dumpable
      * Creates a new ArrayByteBufferPool with a default configuration.
-     * Both {@code maxHeapMemory} and {@code maxDirectMemory} default to 0 to use default heuristic.
+     * Both {@code maxHeapMemory} and {@code maxDirectMemory} default to 0 to use default heuristic;
+     * pooled buffers are hard-referenced at all times.
```
No need to specify `hard-referenced`, restore previous javadoc.
Also for all other constructors.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ConcurrentPool.java:
```diff
@@ -449,6 +490,8 @@ private boolean terminate()
         private boolean tryEnable(boolean acquire)
         {
+            if (!acquire)
+                getHolder().hold();
```
In `tryAcquire()`, the call to `free()` is done after the CAS. I think we should build the `Holder` with a strong reference by default, and only when we successfully CAS to 1 we call `free()`.
In this way the two behaviors (here and in `tryAcquire()`) are normalized.

gregw replied:
@sbordet Your change was a little different to this comment, but I think it is good.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ConcurrentPool.java:
```diff
@@ -481,7 +524,15 @@ private boolean tryAcquire()
+                    if (newMultiplexCount == 1)
+                    {
+                        // We have acquired the entry for the first time, but might be racing with a previous release
+                        // to hold it, so we spin wait to ensure it is held before we free it.
```
This comment should be moved inside `free()` because here is there is no loop, so it's not clear what the comment refers to.

sbordet committed:
* Avoid using deprecated ConcurrentPool constructors.
* Removed unused boolean parameter "weakPool" in ArrayByteBufferPool and updated javadocs.
* Improved implementation of ConcurrentPool.Entry.tryEnable().
* Added javadocs for ConcurrentPool.Holder.free() and hold().

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Merged branch 'jetty-12.0.x' into 'experiment/jetty-12/WeakReference-ConcurrentPool'.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>
